src/main/java/com/shapesecurity/functional/data/ImmutableSet.java

```diff
@@ -62,6 +62,12 @@ public boolean contains(@Nonnull T datum) {
         return this.foldAbelian((val, acc) -> acc.put(f.apply(val)), ImmutableSet.empty((Hasher<A>) this.data.hasher));
     }
 
+    @Nonnull
+    @SuppressWarnings("unchecked")
+    public <A> ImmutableSet<A> flatMap(@Nonnull F<T, ImmutableSet<A>> f) {
```

Comment:
Not sure how I feel about this one. I would be happier if we checked that the `hasher`s were all the same, but that's usually going to be wasted effort. And I'd also be happier if `F` returned an arbitrary `Iterable`, but Java doesn't understand variance in generics.

I'm not sure this comes up enough to be worth it.

Reply:
Mmm, I can see some very odd and a pain to debug bugs coming about from not checking the hashers, so I think it's worth it. I'm not worried about the `Iterable` case here, it would be slower than `union`ing sets which will almost always be the case I think.